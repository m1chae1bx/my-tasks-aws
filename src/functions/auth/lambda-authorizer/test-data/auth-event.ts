import { baseTokenAuthorizerEvent } from "@test-utils/base-objects";

export const authEvent = baseTokenAuthorizerEvent;
