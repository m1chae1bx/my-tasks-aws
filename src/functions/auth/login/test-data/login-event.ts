import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";

export const loginEvent = {
  ...baseAPIGatewayProxyEvent,
  body: JSON.stringify({
    id: "testUserId",
    password: "secret-password",
  }),
};
