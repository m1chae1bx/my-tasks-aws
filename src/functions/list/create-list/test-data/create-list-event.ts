import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";

export const createListEvent = {
  ...baseAPIGatewayProxyEvent,
  pathParameters: {
    userId: "testUserId",
  },
  body: JSON.stringify({
    name: "testListName",
    isDefault: false,
  }),
};
