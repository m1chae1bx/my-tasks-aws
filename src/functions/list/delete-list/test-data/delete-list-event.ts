import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";

export const deleteListEvent = {
  ...baseAPIGatewayProxyEvent,
  pathParameters: {
    userId: "testUserId",
    listId: "testListId",
  },
};
