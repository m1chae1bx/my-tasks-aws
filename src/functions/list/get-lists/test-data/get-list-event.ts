import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";

export const getListEvent = {
  ...baseAPIGatewayProxyEvent,
  pathParameters: {
    userId: "testUserId",
  },
};
