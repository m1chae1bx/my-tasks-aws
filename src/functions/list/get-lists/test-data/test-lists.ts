import { List } from "/opt/nodejs/list.model";

export const testLists: Partial<List>[] = [
  {
    id: "dfab1ffd-d3f3-4c45-b39a-46fcfd32fe7d",
    name: "List 1",
    userId: "testUserId",
    isDefault: false,
  },
  {
    id: "7a7a528c-142f-4b77-a168-8d6b8b56e40e",
    name: "List 2",
    userId: "testUserId",
    isDefault: false,
  },
];
