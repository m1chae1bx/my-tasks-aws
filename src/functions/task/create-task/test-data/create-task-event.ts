import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";

export const createTaskEvent = {
  ...baseAPIGatewayProxyEvent,
  pathParameters: {
    listId: "test-list-id",
  },
  body: JSON.stringify({
    name: "test-task-name",
    desc: "test-task-desc",
    dueDate: "2020-01-01T00:00:00.000Z",
  }),
};
