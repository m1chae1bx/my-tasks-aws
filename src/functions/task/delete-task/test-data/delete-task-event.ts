import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";

export const deleteTaskEvent = {
  ...baseAPIGatewayProxyEvent,
  pathParameters: {
    listId: "test-list-id",
    taskId: "test-task-id",
  },
};
