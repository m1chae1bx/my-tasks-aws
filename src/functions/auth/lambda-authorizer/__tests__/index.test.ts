import { APIGatewayAuthorizerResult } from "aws-lambda";
import { handler } from "../index";
import { authEvent } from "../test-data/auth-event";
import jwt from "jsonwebtoken";

jest.mock("jsonwebtoken", () => ({
  verify: jest.fn(),
}));

describe("lambda-authorizer", () => {
  beforeAll(() => {
    jest.spyOn(console, "error").mockImplementation(() => {});
    jest.spyOn(console, "warn").mockImplementation(() => {});
  });

  describe("happy path", () => {
    let result: APIGatewayAuthorizerResult;

    beforeAll(async () => {
      process.env = {
        AWS_REGION: "test-region",
        ACCOUNT_ID: "test-account-id",
        AWS_LAMBDA_FUNCTION_NAME: "test-lambda-function-name",
        JWT_SECRET: "test-jwt-secret",
      };
      (jwt.verify as jest.Mock).mockReturnValueOnce({
        id: "test-user-id",
        email: "test@email.com",
      });

      result = await handler(authEvent);
    });

    it("should return allow IAM policy", () => {
      expect(result.policyDocument.Statement[0]).toEqual({
        Action: "execute-api:Invoke",
        Effect: "Allow",
        Resource: `arn:aws:execute-api:${process.env.AWS_REGION}:${process.env.ACCOUNT_ID}:*`,
      });
      expect(result.principalId).toBe("test-user-id");
      expect(result.context?.email).toEqual("test@email.com");
    });
  });

  describe("sad path - no JWT_SECRET", () => {
    let err: unknown;

    beforeAll(async () => {
      process.env = {
        AWS_REGION: "test-region",
        ACCOUNT_ID: "test-account-id",
        AWS_LAMBDA_FUNCTION_NAME: "test-lambda-function-name",
      };
      try {
        await handler(authEvent);
      } catch (error) {
        err = error;
      }
    });

    it("should return an error", () => {
      expect(err).toMatchSnapshot();
    });
  });

  describe("sad path - error verifying token", () => {
    let result: APIGatewayAuthorizerResult;

    beforeAll(async () => {
      jest.spyOn(console, "error").mockClear();
      process.env = {
        AWS_REGION: "test-region",
        ACCOUNT_ID: "test-account-id",
        AWS_LAMBDA_FUNCTION_NAME: "test-lambda-function-name",
        JWT_SECRET: "test-jwt-secret",
      };
      (jwt.verify as jest.Mock).mockImplementationOnce(() => {
        throw new Error("test-error");
      });
      result = await handler({ ...authEvent, authorizationToken: "invalid" });
    });

    it("should log error", () => {
      expect(jest.spyOn(console, "error").mock.calls).toMatchSnapshot();
    });
    it("should return deny IAM policy", () => {
      expect(result.policyDocument.Statement[0]).toEqual({
        Action: "execute-api:Invoke",
        Effect: "Deny",
        Resource: `arn:aws:execute-api:${process.env.AWS_REGION}:${process.env.ACCOUNT_ID}:*`,
      });
    });
  });

  describe("sad path - JWT payload is not UserDetailsPayload", () => {
    let result: APIGatewayAuthorizerResult;

    beforeAll(async () => {
      jest.spyOn(console, "warn").mockClear();
      process.env = {
        AWS_REGION: "test-region",
        ACCOUNT_ID: "test-account-id",
        AWS_LAMBDA_FUNCTION_NAME: "test-lambda-function-name",
        JWT_SECRET: "test-jwt-secret",
      };
      (jwt.verify as jest.Mock).mockReturnValueOnce({});
      result = await handler(authEvent);
    });

    it("should log warning", () => {
      expect(jest.spyOn(console, "warn").mock.calls).toMatchSnapshot();
    });

    it("should return allow IAM policy", () => {
      expect(result.policyDocument.Statement[0]).toEqual({
        Action: "execute-api:Invoke",
        Effect: "Allow",
        Resource: `arn:aws:execute-api:${process.env.AWS_REGION}:${process.env.ACCOUNT_ID}:*`,
      });
    });

    it("should return empty principalId and context", () => {
      expect(result.principalId).toBe("");
      expect(result.context).toBe(undefined);
    });
  });
});
