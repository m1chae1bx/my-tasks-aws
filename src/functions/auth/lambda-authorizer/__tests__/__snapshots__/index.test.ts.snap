// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lambda-authorizer sad path - JWT payload is not UserDetailsPayload should log warning 1`] = `
Array [
  Array [
    "JWT payload is not a UserDetailsPayload",
  ],
]
`;

exports[`lambda-authorizer sad path - error verifying token should log error 1`] = `
Array [
  Array [
    "Error verifying token",
    [Error: test-error],
  ],
]
`;

exports[`lambda-authorizer sad path - no JWT_SECRET should return an error 1`] = `[Error: JWT_SECRET environment variable is not set in Lambda function test-lambda-function-name]`;
